### PR TITLE
New state system

### PR DESCRIPTION
**Pull request by jonas747**
This is a draft pr tracking my progress on the new state system.

There's probably still a couple things i will change on the design but the higher level structure of it is set.

As for why there's a new state system can be found here: https://github.com/jonas747/dstate/blob/v3/README.md

### CODE REVIEW

File: bot/botrest/server.go [hunk 1]
```diff
@@ -217,15 +204,19 @@ func HandleChannelPermissions(w http.ResponseWriter, r *http.Request) {
...
 	if guild == nil {
 		internalapi.ServerError(w, r, errors.New("Guild not found"))
 		return
 	}
 
-	perms, err := guild.MemberPermissions(true, cId, common.BotUser.ID)
-
+	member, err := bot.GetMember(gId, common.BotUser.ID)
 	if err != nil {
+		internalapi.ServerError(w, r, errors.New("Could not find bot member"))
```
Comment (jo3-l) [inline bot/botrest/server.go:215]: should there be a `return` after this ?

Comment (l-zeuch) [reply bot/botrest/server.go:215]: The professional nitpicker is back at his workplace 👀

**Comment by jonas747** [comment]
This is getting close, there's just a small number of tasks left:

 - Update shard migration to support the new state system (might require a small modifcation to dstate.InMemoryTracker)
 - Add members to cache from the member fetcher (might require a small modifcation to dstate.InMemoryTracker)
 - Ensure backwards compat with custom commands/templates
 - testing testing testing

**Review by jonas747 — COMMENTED** [review]
finally done going over 147 files

**Pull request merged**